app/src/block_hash.rs

```diff
@@ -1,37 +1,411 @@
+                block_number
+            );
+        }
+
+        self.entries.push_back(BlockHashEntry {
+            number: block_number,
+            hash: block_hash,
+        });
+
+        while self.entries.len() > BLOCKHASH_HISTORY_SIZE {
+            self.entries.pop_front();
+        }
+
+        self.latest_block = block_number;
+    }
+
+    pub fn initialize_from_storage<S, B>(storage: &S, block_query: &B) -> Self
+    where
+        B: BlockQueries<Storage = S>,
+    {
+        let latest_block = block_query.latest(storage).unwrap().unwrap_or(0);
+
+        let mut cache = Self::default();
+
+        let start_block = if latest_block >= BLOCKHASH_HISTORY_SIZE as u64 {
+            latest_block - BLOCKHASH_HISTORY_SIZE as u64 + 1
+        } else {
+            0
+        };
+
+        for block_num in start_block..=latest_block {
+            if let Ok(Some(block)) = block_query.by_height(storage, block_num, false) {
+                cache.push(block_num, block.0.header.hash);
+            }
+        }
+
+        cache
+    }
+}
+
+impl Default for BlockHashRingBuffer {
+    fn default() -> Self {
+        Self {
+            entries: VecDeque::with_capacity(BLOCKHASH_HISTORY_SIZE),
+            latest_block: 0,
+        }
+    }
+}
+
+impl BlockHashLookup for BlockHashRingBuffer {
+    fn hash_by_number(&self, block_number: u64) -> Option<B256> {
+        if block_number > self.latest_block {
+            return None;
+        }
+
+        let blocks_ago = (self.latest_block - block_number) as usize;
+
+        // second condition to guard against cases when the buffer is not full
+        // and starts from non-zero block
+        if blocks_ago > BLOCKHASH_HISTORY_SIZE || blocks_ago >= self.entries.len() {
+            return None;
+        }
+
+        let block_index = self.entries.len() - 1 - blocks_ago;
+
+        self.entries.get(block_index).map(|e| e.hash)
+    }
+}
+
+impl BlockHashWriter for BlockHashRingBuffer {
+    fn push(&mut self, height: u64, hash: B256) {
+        self.push(height, hash);
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SharedBlockHashCache {
+    inner: Arc<RwLock<BlockHashRingBuffer>>,
+}
+
+impl SharedBlockHashCache {
+    pub fn initialize_from_storage<S, B>(storage: &S, block_query: &B) -> Self
+    where
+        B: BlockQueries<Storage = S>,
+    {
+        let buf = BlockHashRingBuffer::initialize_from_storage(storage, block_query);
+
+        Self {
+            inner: Arc::new(RwLock::new(buf)),
+        }
+    }
+}
+
+impl Default for SharedBlockHashCache {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(BlockHashRingBuffer::default())),
+        }
+    }
+}
+
+impl BlockHashLookup for SharedBlockHashCache {
+    fn hash_by_number(&self, number: u64) -> Option<B256> {
+        if let Ok(cache) = self.inner.read() {
+            cache.hash_by_number(number)
+        } else {
+            // TODO: deal with poisoning?
```

Comment:
Is this a real issue? If there is no cache set up yet, then there is no block hash in it, right?